Add docs on how we promote stable releases (#13749)
openxla/iree: docs/developers/developing_iree/releasing.md

# Releasing

IREE cuts automated releases via a workflow that is
[triggered daily](https://github.com/openxla/iree/blob/main/.github/workflows/schedule_candidate_release.yml).
The only constraint placed on the commit that is released is that it has passed
all CI checks. These are published on GitHub with the "pre-release" status. For
debugging this process, see the
[Debugging releases playbook](../debugging/releases.md).

We periodically promote one of these candidates to a "stable" release by
removing the "pre-release" status. This makes it show up as a "latest" release
on GitHub. We also push the Python packages for this release to PyPI.

The criteria for selecting this candidate is a bit more involved.

## Coupling to the Google Integrate

The Google team that manages these stable releases at the moment is also
responsible for integrating the IREE source code into Google's monorepo. For
convenience, we select a candidate pre-release, attempt to integrate it into
Google's monorepo and then promote it to stable if that was successful without
cherry picks that would affect the quality of the release (because they wouldn't
be present in the promoted version).

> This gives some additional validation to the release because it is stress-tested
> running in a different environment and we not-infreuqently catch some issues. We

Reviewer: typo, fix or just remove the double negative

```suggestion
This gives some additional validation to the release because it is stress-tested
running in a different environment and we periodically catch some issues. We
```

Author: Hmm I think periodically means something different :grin: 

```suggestion
This gives some additional validation to the release because it is stress-tested
running in a different environment and we not-infrequently catch some issues. We
```

Reviewer: "not infrequently" == "frequently" == "periodically"? IDK... "occasionally", "sometimes", "rarely"?

Author: It's not *infrequent* but that doesn't mean it's *frequent*. I think sometimes is closest, but I think the emphasis is different. "periodically" focuses on it happening at regular intervals which isn't the case here at all. Anyway, I'm going to leave this unless you care a lot :-)

do not currently have a way to add cherry-picks to a release, so if cherry-picks
for functional issues are required, we skip promoting the candidate to "stable".

This coupling introduces some additional constraints to the process that are not
inherent. It would be perfectly fine to start promoting candidates based on some
other process, but since the same people are managing both, we've coupled these
so we don't have to keep track of as many different versions.

As the project matures, we will likely remove this coupling. In particular we
will likely start integrating into Google's monorepo at a faster cadence than
the stable releases, so a 1:1 mapping there will not make sense.

The PyPI password is also currently stored in Google's internal secret
management system, so for others to manage the deployment, we would need to

> store it elsewhere withe appropriate ACLs.

Reviewer: typo
```suggestion
store it elsewhere with appropriate ACLs.
```

At the point where others want to engage in the release process, we can easily
remove any coupling to any Google processes.

## Picking a Candidate to Promote

When selecting a candidate we use the following criteria:

1. ⪆4 days old so that problems with it may have been spotted.
2. Contains no P0 regressions vs the previous stable release.
3. LLVM submodule commit exists upstream (no cherry picks or patches) and
   matches a commit already integrated into Google's monorepo

The constraint on LLVM version is largely due to our current process for doing
so. We aim to lift this limitation and if the process were decoupled from the
Google integration (see
[Coupling to the Google Integrate](#coupling-to-the-google-integrate)), it would
go away anyway.

There is currently no specific tracking for P0 regressions (process creation in
progress). When you've identified a potential candidate, email the iree-discuss
list with the proposal and solicit feedback. People may point out known
regressions or request that some feature make the cut.

## Releasing

1. (Googlers only) Integrate into Google's monorepo, following
   http://go/iree-g3-integrate-playbook.

> 2. Open the release on GitHub. Rename the release from “candidate to “stable",

Reviewer 2: "candidate candidate" to "stable stable" (missing a quote mark and I think here and in the picture it makes sense to write it twice as that's what we've been doing - https://github.com/openxla/iree/releases?q=%22stable+stable%22&expanded=true. unless we want to change this.

Author: Yeah I think we do change both right now, but I don't think that actually makes sense. The tag is still candidate-*, so changing that bit is confusing IMO. @ScottTodd thoughts? Do you remember why we were renaming both?

Reviewer: Ehhh, chatted with Geoffrey. `stable candidate-[date].[index]` sort of makes sense, considering the git tag is `candiate-[date].[index]`

   uncheck the option for “pre-release”, and check the option for "latest".

   ![rename_release](../assets/rename_release.png)

   ![promote_release](../assets/promote_release.png)

> 3. (Googlers only) Push to PyPI using

Reviewer 2: Also should running the script come before renaming? Just thinking of this last release that took me >day to figure out how to correct the deploy script :)

Author: Hmm yeah probably. Ideally they would actually be atomic, but I guess renaming is unlikely to fail. We should probably script that as well, actually.

Author: Done (the reordering bit, not the scripting bit)
